query.go

```diff
@@ -475,7 +555,23 @@ func (q *query) updateState(ctx context.Context, up *queryUpdate) {
 		if p == q.dht.self { // don't add self.
 			continue
 		}
-		q.queryPeers.TryAdd(p, up.cause)
+
+		if q.dFilter == nil {
+			q.queryPeers.TryAdd(p, up.cause)
+			continue
+		}
+
+		// add the peer as heard if it's a diverse peer.
+		if q.dFilter.TryAdd(p) {
+			// filter allowed the peer. Try to add it to the query set and remove it from
+			// the filter if it's not added to the query set.
+			if !q.queryPeers.TryAdd(p, up.cause) {
+				q.dFilter.Remove(p)
+			}
+		} else {
+			// filter didn't allow the peer. So, add it to the set of query peers rejected by the diversity filter.
+			q.dfRejectedPeers.TryAdd(p, up.cause)
+		}
```

Comment:
There's an issue here where since we mark peers as non-diverse when they are added to the queue, not popped off the queue this means that if I hear about `maxAllowed + 1` (e.g. 3) peers in the same IP range that only `maxAllowed` (e.g. 2) of them are actually queryable. This seems like a potential attack since an attacker can return a set of peers from IP ranges/ASNs that it wants to **deny** the client access to without actually owning any peers in that range (e.g. I can just make up an AWS IP range and when you get a dial failure that'll be enough to stop you from querying other AWS peers).

Figuring this out is a bit tricky because of the asynchronous nature of things. One thought I had is if we're willing to potentially waste some dials maybe we could add something like this:
1. Add `Rejected` state to the peer queue
2. In the `updateState` function when we receive a response see if the peer is allowed and if not mark it as `Rejected`, take the `queryUpdate` information and stash it in some map of rejected peers that's on the query instance
3. Have some strategy for minimizing the number of peers we query that end up `Rejected` (e.g. when processing a new peer that is entering the `Queried` state go through the peer queue and mark disallowed peers as `Rejected`. Alternatively, we could add another check in `spawnQuery` for rejecting peers and process those rejections in `updateState`).

Reply:
There may also be some advantages to be gained from limiting the number of outbound queries to a single IP range so as to limit the issues above:

For example, in `spawnQuery` we could select the top candidate that is not in the same IP range as one of the peers that is in the `Queried` state. This could potentially prevent us from having to discard a peer due to filter rejection. However, since we're not changing the termination condition this "skip" is just a temporary measure. Of course, if a peer should actually be rejected (since we have enough successful peers in the IP range), then we should reject it and not check it again.

Question: This still allows peer A to effect when peer B is queried by delaying the query. I don't think this matters though given the constraints we're imposing on the attackers in this scenario. lmk if you disagree.